clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,347 @@
+            }
+        }
+    }
+
+    protected void collectMetric(MetricsEmitter metricsEmitter, MetricKey metricKey, KafkaMetric metric) {
+        Object metricValue;
+
+        try {
+            metricValue = metric.metricValue();
+        } catch (Exception e) {
+            // If an exception occurs when retrieving value, log warning and continue to process the rest of metrics
+            log.warn("Failed to retrieve metric value {}", metricKey.getName(), e);
+            return;
+        }
+
+        if (isMeasurable(metric)) {
+            Measurable measurable = metric.measurable();
+            Double value = (Double) metricValue;
+
+            if (measurable instanceof WindowedCount || measurable instanceof CumulativeSum) {
+                collectSum(metricKey, value, metricsEmitter);
+                collectDelta(metricKey, value, metricsEmitter);
+            } else {
+                collectGauge(metricKey, value, metricsEmitter);
+            }
+        } else {
+            // It is non-measurable Gauge metric.
+            // Collect the metric only if its value is a number.
+            if (metricValue instanceof Number) {
+                Number value = (Number) metricValue;
+                collectGauge(metricKey, value, metricsEmitter);
+            } else {
+                // skip non-measurable metrics
+                log.debug("Skipping non-measurable gauge metric {}", metricKey.getName());
+            }
+        }
+    }
+
+    private void collectDelta(MetricKey originalKey, Double value, MetricsEmitter metricsEmitter) {
+        MetricKey metricKey = metricNamingStrategy.derivedMetricKey(originalKey, "delta");
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        // calculate a getAndSet, and add to out if non-empty
+        final Instant timestamp = clock.instant();
+        InstantAndValue<Double> instantAndValue = ledger.delta(originalKey, timestamp, value);
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.deltaSum(metricKey, instantAndValue.getValue(), true, timestamp,
+                instantAndValue.getIntervalStart())
+        );
+    }
+
+    private void collectSum(MetricKey metricKey, double value, MetricsEmitter metricsEmitter) {
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.sum(metricKey, value, true, clock.instant(), ledger.instantAdded(metricKey))
+        );
+    }
+
+    private void collectGauge(MetricKey metricKey, Number value, MetricsEmitter metricsEmitter) {
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.gauge(metricKey, value, clock.instant())
+        );
+    }
+
+    @Override
+    public String toString() {
+        return this.getClass().getCanonicalName();
```

Comment:
Any particular reason why we overwrite `toString()` -- seems not to expose any internal information, so no sure what the purpose is?

Reply:
Yeah can be removed, I don't think it has much significance. Done.